README.md
# Archiver v0.9.1 [![Build Status](https://travis-ci.org/ctalkington/node-archiver.svg?branch=master)](https://travis-ci.org/ctalkington/node-archiver)

a streaming interface for archive generation

[![NPM](https://nodei.co/npm/archiver.png)](https://nodei.co/npm/archiver/)

## Install

```bash
npm install archiver --save
```

You can also use `npm install https://github.com/ctalkington/node-archiver/archive/master.tar.gz` to test upcoming versions.

## Archiver

#### create(format, options)

Creates an Archiver instance based on the format (zip, tar, etc) passed. Parameters can be passed directly to `Archiver` constructor for convenience.

#### registerFormat(format, module)

Registers an archive format. Format modules are essentially transform streams with a few required methods. They will be further documented once a formal spec is in place.

### Instance Methods

Inherits [Transform Stream](http://nodejs.org/api/stream.html#stream_class_stream_transform) methods.

#### append(input, data)

Appends an input source (text string, buffer, or stream) to the instance. When the instance has received, processed, and emitted the input, the `entry` event is fired.

Replaced `#addFile` in v0.5.

```js
archive.append('string', { name:'string.txt' });
archive.append(new Buffer('string'), { name:'buffer.txt' });
archive.append(fs.createReadStream('mydir/file.txt'), { name:'stream.txt' });
archive.append(null, { name:'dir/' });
```

#### bulk(mappings)

Appends multiple entries from passed array of src-dest mappings. A lazystream wrapper is used to prevent issues with open file limits.

Globbing patterns are supported through use of the [file-utils](https://github.com/SBoudrias/file-utils) package. Please note that multiple src files to single dest file (ie concat) is not supported.

The `data` property can be set (per src-dest mapping) to define data for matched entries.

```js
archive.bulk([
  { src: ['mydir/**'], data: { date: new Date() } },
  { expand: true, cwd: 'mydir', src: ['**'], dest: 'newdir' }
]);
```

For more detail on this feature, please see [BULK.md](https://github.com/ctalkington/node-archiver/blob/master/BULK.md).

#### file(filepath, data)

Appends a file given its filepath using a lazystream wrapper to prevent issues with open file limits. When the instance has received, processed, and emitted the file, the `entry` event is fired.

```js
archive.file('mydir/file.txt', { name:'file.txt' });
```

#### finalize()

Finalizes the instance. You should listen for the `end`/`close`/`finish` of the destination stream to properly detect completion.

#### pointer()

Returns the current byte length emitted by archiver. Use this in your end callback to log generated size.

## Events

Inherits [Transform Stream](http://nodejs.org/api/stream.html#stream_class_stream_transform) events.

#### entry

Fired when the input has been received, processed, and emitted. Passes entry data as first argument.

## Zip

### Options

#### comment `string`

Sets the zip comment.

#### forceUTC `boolean`

If true, forces the entry date to UTC. Helps with testing across timezones.

#### store `boolean`

If true, all entry contents will be archived without compression by default.

#### zlib `object`

Passed to node's [zlib](http://nodejs.org/api/zlib.html#zlib_options) module to control compression. Options may vary by node version.

### Entry Data

#### name `string` `required`

Sets the entry name including internal path.

#### date `string|Date`

Sets the entry date. This can be any valid date string or instance. Defaults to current time in locale.

#### store `boolean`

If true, entry contents will be archived without compression.

#### comment `string`

Sets the entry comment.

#### mode `number`

Sets the entry permissions. Defaults to octal 0755 (directory) or 0644 (file).

## Tar

### Options

#### gzip `boolean`

Compresses the tar archive using gzip, default is false.

#### gzipOptions `object`

Passed to node's [zlib](http://nodejs.org/api/zlib.html#zlib_options) module to control compression. Options may vary by node version.

### Entry Data

#### name `string` `required`

Sets the entry name including internal path.

#### date `string|Date`

Sets the entry date. This can be any valid date string or instance. Defaults to current time in locale.

#### mode `number`

Sets the entry permissions. Defaults to octal 0755 (directory) or 0644 (file).

## Libraries

Archiver makes use of several libraries/modules to avoid duplication of efforts.

- [zip-stream](https://npmjs.org/package/zip-stream)
- [tar-stream](https://npmjs.org/package/tar-stream)

## Things of Interest

- [Examples](https://github.com/ctalkington/node-archiver/blob/master/examples)
- [Changelog](https://github.com/ctalkington/node-archiver/releases)
- [Contributing](https://github.com/ctalkington/node-archiver/blob/master/CONTRIBUTING.md)
- [MIT License](https://github.com/ctalkington/node-archiver/blob/master/LICENSE-MIT)

CHANGELOG
v0.9.1:
  date: 2014-05-04
  changes:
    - misc cleanup. see commit log.
v0.9.0:
  date: 2014-04-18
  changes:
    - restrict files published to npm. (GH #71)
    - replace implementation of tar with tar-stream to avoid duplication of efforts. (GH #63)
    - always use readable-stream for consistency.
    - more functionality tests. (more format specific ones coming soon)
    - updated documentation and examples.
    - other misc cleanup and tweaks. see git log.
v0.8.1:
  date: 2014-04-01
  changes:
    - core: fix edge case in bulk() where the dot representing the current directory would be included.
v0.8.0:
  date: 2014-04-01
  changes:
    - initial directory entry support.
    - normalize null sources to empty buffer.
    - crc32: replace with buffer-crc32.
    - other small tweaks. see git log.
v0.7.1:
  date: 2014-03-27
  changes:
    - core: catch module errors and bubble them up the chain.
    - core: create new extended data object in bulk loop to prevent data corruption. (GH #67)
    - tar: catch compressor errors and bubble them up the chain.
    - zip: pass `on` calls to underlying engine.
    - other small tweaks. see git log.
v0.7.0:
  date: 2014-03-22
  changes:
    - tar: gzip helper (GH #65)
v0.6.1:
  date: 2014-02-16
  changes:
    - bugfixes and small tweaks. see git log.
v0.6.0:
  date: 2014-02-15
  changes:
    - core: switch to events fully.
    - core: internal var cleanup.
    - core: highWaterMark to 1MB.
    - core: add pointer() to get emitted byte length.
    - core: remove early exit detection as its served its initial purpose.
    - zip: global store option.
v0.5.2:
  date: 2014-01-24
  changes:
    - core: add experimental catchEarly option (GH #59)
    - core: drop readable-stream to node v0.10 compatible version
v0.5.1:
  date: 2014-01-15
  changes:
    - adjust when module output is piped. (GH #57)
v0.5.0:
  date: 2014-01-10
  changes:
    - removal of node v0.6 support.
    - new file() and bulk() helpers.
    - new json format module for testing.
    - experimental zip file mode support.
    - massive overhaul to become more modular.
    - removal of addFile.
    - and much more. check out git log.
v0.4.10:
  date: 2013-09-28
  changes:
    - tar: fix handling of numeric values. (GH #45)
    - formats: add format specs.
    - misc doc and test refactoring.
v0.4.9:
  date: 2013-09-02
  changes:
    - tar: zero length checks (GH #44)
    - core: more queue handling tweaks.
    - misc code and test refactoring.
v0.4.8:
  date: 2013-08-18
  changes:
    - iconv-lite is more than a devDep. (GH #42)
v0.4.7:
  date: 2013-08-18
  changes:
    - core: bit of queue handling tweaks.
    - core: increase internal buffer size.
    - zip: always convert filenames to utf8. (GH #41)
    - zip: workaround race condition with deflate.pipe. thanks @inolen (GH #32)
v0.4.6:
  date: 2013-07-07
  changes:
    - core: introduce _normalizeFileData internally to seperate logic a bit.
    - headers: internal api updates.
    - tar: use prefix when name is > 100 and <= 255 (GH #35)
v0.4.5:
  date: 2013-06-25
  changes:
    - force STORE when compression level is zero. (GH #33)
v0.4.4:
  date: 2013-06-08
  changes:
    - switch test suite to mocha. test overhaul.
    - support streams without any buffer or pause method
v0.4.3:
  date: 2013-04-17
  changes:
    - default highWaterMark to 64kb.
    - attempt to catch when process exits before it should.
    - code cleanup.
v0.4.2:
  date: 2013-04-09
  changes:
    - rebased on Transform stream for internal pipe-ing.
    - tweaks to internal header api.
    - pause/resume only called when _ReadableState isn't available (ie node < 0.10).
v0.4.1:
  date: 2013-03-13
  changes:
    - doc fixes.
v0.4.0:
  date: 2013-02-16
  changes:
    - basic benchmarks and optimizations. props @danmilon. (GH #20,21)
    - misc optimizations. props @visionmedia. (GH #17,19)
    - removal of lodash dependency.
    - massive overhaul to processing flow, methods are async now too.
    - use readable-stream with streams2-like api.
    - fix examples and make runnable out of box. props @danmilon. (GH #11)
v0.3.0:
  date: 2013-01-27
  changes:
    - use setImmediate when it exists. props @davglass. (GH #9)
    - emit proper error objects throughout.
    - replace error emitting on methods with callback.
    - add generic create method.
    - make tar recordSize configurable for advanced users.
v0.2.2:
  date: 2013-01-05
  changes:
    - fix tar stream bug and add stream tests.
v0.2.1:
  date: 2013-01-04
  changes:
    - refactor tests and use forceUTC so we can test with date objects.
    - add forceUTC option. helps with testing across timezones.
    - bugfix: zip date was using invalid function.
v0.2.0:
  date: 2013-01-02
  changes:
    - Add ability to archive with Tar (or TGZ with little effort).
    - Reorganization to make further updates smaller and focused.
v0.1.1:
  date: 2012-11-14
  changes:
    - Break core and Zip logic apart.
    - Fix sub-directories parsing issue with some Zip applications.
v0.1.0:
  date: 2012-10-09
  changes:
    - First release.
    - Import ZipStream logic.

lib/archive.js
/*
 * node-archive
 *
 * Copyright (c) 2012 Chris Talkington, contributors.
 * Licensed under the MIT license.
 * https://github.com/ctalkington/node-archive/blob/master/LICENSE-MIT
 */

var stream = require('stream');
var util = require('util');
var zlib = require('zlib');

var crc32 = require('./util/crc32');

function Archive(opt) {
  var self = this;

  self.readable = true;
  self.paused = false;
  self.busy = false;
  self.eof = false;

  self.queue = [];
  self.fileptr = 0;
  self.files = [];
  self.options = opt;
}

util.inherits(Archive, stream.Stream);

exports.createZip = function(opt) {
  return new Archive(opt);
};

// converts datetime to DOS format
function convertDate(d) {
  var year = d.getFullYear();

  if (year < 1980) {
    return (1<<21) | (1<<16);
  }
  return ((year-1980) << 25) | ((d.getMonth()+1) << 21) | (d.getDate() << 16) |
    (d.getHours() << 11) | (d.getMinutes() << 5) | (d.getSeconds() >> 1);
}


Archive.prototype.pause = function() {
  var self = this;
  self.paused = true;
};

Archive.prototype.resume = function() {
  var self = this;
  self.paused = false;

  self._read();
};

Archive.prototype.destroy = function() {
  var self = this;
  self.readable = false;
};

Archive.prototype._read = function() {
  var self = this;

  if (!self.readable || self.paused) { return; }

  if (self.queue.length > 0) {
    var data = self.queue.shift();
    self.emit('data', data);
  }

  if (self.eof && self.queue.length === 0) {
    self.emit('end');
    self.readable = false;

    if (self.callback) {
      self.callback(self.fileptr);
    }
  }

  process.nextTick(function() { self._read(); }); //TODO improve
};

Archive.prototype.finalize = function(callback) {
  var self = this;

  if (self.files.length === 0) {
    self.emit('error', 'no files in zip');
    return;
  }

  self.callback = callback;
  self._pushCentralDirectory();
  self.eof = true;
};

Archive.prototype._addFileStore = function(source, file, callback) {
  // placeholder
};

Archive.prototype._addFileDeflate = function(source, file, callback) {
  // placeholder
};

Archive.prototype.addFile = function(source, file, callback) {
  var self = this;

  if (self.busy) {
    self.emit('error', 'previous file not finished');
    return;
  }

  if (typeof source === 'string') {
    source = new Buffer(source, 'utf-8');
  }

  self.busy = true;
  self.file = file;
  self._pushLocalFileHeader(file);

  var checksum = crc32.createCRC32();
  file.uncompressed = 0;
  file.compressed = 0;

  function onEnd() {
    file.crc32 = checksum.digest();
    if (file.store) { file.compressed = file.uncompressed; }

    self.fileptr += file.compressed;
    self._pushDataDescriptor(file);

    self.files.push(file);
    self.busy = false;
    callback();
  }

  function update(chunk) {
    checksum.update(chunk);
    file.uncompressed += chunk.length;
  }

  if (file.store) {
    if (Buffer.isBuffer(source)) {
      update(source);

      self.queue.push(source);
      process.nextTick(onEnd);
    } else {
      // Assume stream
      source.on('data', function(chunk) {
        update(chunk);
        self.queue.push(chunk);
      });

      source.on('end', onEnd);
    }
  } else {
    var deflate = zlib.createDeflateRaw(self.options);

    deflate.on('data', function(chunk) {
      file.compressed += chunk.length;
      self.queue.push(chunk);
    });

    deflate.on('end', onEnd);

    if (Buffer.isBuffer(source)) {
      update(source);
      deflate.write(source);
      deflate.end();
    } else {
      // Assume stream
      source.on('data', function(chunk) {
        update(chunk);
        deflate.write(chunk); //TODO check for false & wait for drain
      });
      source.on('end', function() {
        deflate.end();
      });
    }
  }

  process.nextTick(function() { self._read(); });
};

// TODO remove listeners on end

// local file header
Archive.prototype._pushLocalFileHeader = function(file) {
  var self = this;

  file.version = 20;
  file.bitflag = (1<<3) | (1<<11);
  file.method = file.store ? 0 : 8;
  if (!file.date) { file.date = new Date(); }
  file.moddate = convertDate(file.date);
  file.offset = self.fileptr;

  var buf = new Buffer(1024);
  var len;

  buf.writeUInt32LE(0x04034b50, 0);         // local file header signature
  buf.writeUInt16LE(file.version, 4);       // version needed to extract
  buf.writeUInt16LE(file.bitflag, 6);       // general purpose bit flag
  buf.writeUInt16LE(file.method, 8);        // compression method
  buf.writeUInt32LE(file.moddate, 10);      // last mod file date and time
  buf.writeInt32LE(0, 14);                  // crc32
  buf.writeUInt32LE(0, 18);                 // compressed size
  buf.writeUInt32LE(0, 22);                 // uncompressed size

  buf.writeUInt16LE(0, 28);                 // extra field length
  len = buf.write(file.name, 30);           // file name
  buf.writeUInt16LE(len, 26);               // file name length

  len += 30;
  self.queue.push(buf.slice(0, len));
  self.fileptr += len;
};

Archive.prototype._pushDataDescriptor = function(file) {
  var self = this;

  var buf = new Buffer(16);
  buf.writeUInt32LE(0x08074b50, 0);         // data descriptor record signature
  buf.writeInt32LE(file.crc32, 4);          // crc-32
  buf.writeUInt32LE(file.compressed, 8);    // compressed size
  buf.writeUInt32LE(file.uncompressed, 12); // uncompressed size

  self.queue.push(buf);
  self.fileptr += buf.length;
};

Archive.prototype._pushCentralDirectory = function() {
  var self = this;
  var cdoffset = self.fileptr;

  var ptr = 0;
  var cdsize = 0;

  var len, buf;

  for (var i=0; i<self.files.length; i++) {
    var file = self.files[i];

    buf = new Buffer(1024);

    // central directory file header
    buf.writeUInt32LE(0x02014b50, 0);         // central file header signature
    buf.writeUInt16LE(file.version, 4);       // TODO version made by
    buf.writeUInt16LE(file.version, 6);       // version needed to extract
    buf.writeUInt16LE(file.bitflag, 8);       // general purpose bit flag
    buf.writeUInt16LE(file.method, 10);       // compression method
    buf.writeUInt32LE(file.moddate, 12);      // last mod file time and date
    buf.writeInt32LE(file.crc32, 16);         // crc-32
    buf.writeUInt32LE(file.compressed, 20);   // compressed size
    buf.writeUInt32LE(file.uncompressed, 24); // uncompressed size

    buf.writeUInt16LE(0, 30);                 // extra field length
    buf.writeUInt16LE(0, 32);                 // file comment length
    buf.writeUInt16LE(0, 34);                 // disk number where file starts
    buf.writeUInt16LE(0, 36);                 // internal file attributes
    buf.writeUInt32LE(0, 38);                 // external file attributes
    buf.writeUInt32LE(file.offset, 42);       // relative offset
    len = buf.write(file.name, 46);           // file name
    buf.writeUInt16LE(len, 28);               // file name length

    len += 46;
    ptr = ptr + len;
    self.queue.push(buf.slice(0, len));
  }

  cdsize = ptr;

  // end of central directory record
  len = 22;
  buf = new Buffer(len);

  buf.writeUInt32LE(0x06054b50, 0);           // end of central dir signature
  buf.writeUInt16LE(0, 4);                    // number of this disk
  buf.writeUInt16LE(0, 6);                    // disk where central directory starts
  buf.writeUInt16LE(self.files.length, 8);    // number of central directory records on this disk
  buf.writeUInt16LE(self.files.length, 10);   // total number of central directory records
  buf.writeUInt32LE(cdsize, 12);              // size of central directory in bytes
  buf.writeUInt32LE(cdoffset, 16);            // offset of start of central directory, relative to start of archive
  buf.writeUInt16LE(0, 20);                   // comment length

  ptr = ptr + len;

  self.queue.push(buf);
  self.fileptr += ptr;
};